src/stage_4.rs

```diff
@@ -0,0 +1,343 @@
+use serde::Serialize;
+
+use crate::{
+    stage_3::{DelimitingModifier, NorgASTFlat, ParagraphSegment},
+    CarryoverTag, DetachedModifierExtension, NestableDetachedModifier, ParagraphSegmentToken,
+    RangeableDetachedModifier,
+};
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash, Serialize)]
+pub enum NorgAST {
+    Paragraph(Vec<ParagraphSegment>),
+    NestableDetachedModifier {
+        modifier_type: NestableDetachedModifier,
+        level: u16,
+        extensions: Vec<DetachedModifierExtension>,
+        text: Box<NorgASTFlat>,
+        content: Vec<Self>,
```

Comment:
suggestion: might it make more sense to take this the extra mile and only have a single `content` variable? The text would then be a `Paragraph`, and any children would come afterwards.

I find the differentiation of `text` or `content` a bit vague

Reply:
I originally took the same approach as headings. I can see how it might be a little weird though, and it's probably unnecessary.

Reply:
Can we change the type of `NorgASTFlat::NestableDetachedModifier.content` to match `NorgASTFlat::Heading.title`?

It might be a little messy, but it would make using the result of the parser much nicer, and the types would more accurately reflect reality